src/Datagrid/Pager.php

```diff
@@ -70,7 +70,26 @@ public function computeNbResult()
 
     public function getCurrentPageResults(): iterable
     {
-        return $this->getQuery()->execute();
+        $query = $this->getQuery();
+        if (!$query instanceof ProxyQueryInterface) {
+            throw new \TypeError(sprintf(
+                'The pager query MUST implement %s.',
+                ProxyQueryInterface::class,
+            ));
+        }
+
+        $identifierFieldNames = $query
+            ->getQueryBuilder()
+            ->getEntityManager()
+            ->getMetadataFactory()
+            ->getMetadataFor(current($query->getQueryBuilder()->getRootEntities()))
+            ->getIdentifierFieldNames();
+
+        // Paginator with fetchJoinCollection doesn't work with composite primary keys
+        // https://github.com/doctrine/orm/issues/2910
+        $paginator = new Paginator($query->getDoctrineQuery(), 1 === \count($identifierFieldNames));
```

Comment:
I guess it's unlikely, but `getDoctrineQuery` is not in the interface.

Reply:
Indeed. IMO, we could add a call to `assert($query instanceof ...)` then.

Reply:
Fixed